cmd/collector/app/span_processor.go

```diff
@@ -177,3 +209,67 @@ func (sp *spanProcessor) enqueueSpan(span *model.Span, originalFormat SpanFormat
 	}
 	return sp.queue.Produce(item)
 }
+
+func (sp *spanProcessor) background(reportPeriod time.Duration, callback func()) {
+	go func() {
+		ticker := time.NewTicker(reportPeriod)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				callback()
+			case <-sp.stopCh:
+				return
+			}
+		}
+	}()
+}
+
+func (sp *spanProcessor) updateQueueSize() {
+	if sp.dynQueueSizeWarmup == 0 {
+		return
+	}
+
+	if sp.dynQueueSizeMemory == 0 {
+		sp.logger.Warn("The dynamic queue size warmup value is set, but not the amount of memory to use. Skipping.")
+		return
+	}
+
+	if sp.spansProcessed.Load() < uint64(sp.dynQueueSizeWarmup) {
+		return
+	}
+
+	// first, we get the average size of a span, by dividing the bytes processed by num of spans
+	average := sp.bytesProcessed.Load() / sp.spansProcessed.Load()
+
+	// finally, we divide the available memory by the average size of a span
+	idealQueueSize := float64(sp.dynQueueSizeMemory / uint(average))
+
+	// cap the queue size, just to be safe...
+	if idealQueueSize > maxQueueSize {
+		idealQueueSize = maxQueueSize
+	}
+
+	var diff float64
+	current := float64(sp.queue.Capacity())
+	if idealQueueSize > current {
+		diff = idealQueueSize / current
+	} else {
+		diff = current / idealQueueSize
+	}
+
+	// resizing is a costly operation, we only perform it if we are at least 20% apart from the desired value
+	if diff > 1.2 {
+		s := int(idealQueueSize)
+		sp.logger.Info("Resizing the internal span queue", zap.Int("new-size", s), zap.Uint64("average-span-size-bytes", average))
+		sp.queueResizeMu.Lock()
```

Comment:
I would move the locking higher, before the size calculation starts.

Reply:
Done